sxg_rs/src/wasm_worker.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::headers::AcceptFilter;
+use crate::http::HttpResponse;
+use crate::utils::to_js_error;
+use crate::SxgWorker;
+use anyhow::Result;
+use js_sys::Function as JsFunction;
+use js_sys::Promise as JsPromise;
+use js_sys::Uint8Array;
+use std::sync::Arc;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::future_to_promise;
+
+#[wasm_bindgen]
+pub struct WasmWorker(Arc<SxgWorker>);
+
+#[wasm_bindgen]
+impl WasmWorker {
+    #[wasm_bindgen(constructor)]
+    pub fn new(config_yaml: &str, cert_pem: &str, issuer_pem: &str) -> Result<WasmWorker, JsValue> {
+        let sxg_worker = SxgWorker::new(config_yaml, cert_pem, issuer_pem).map_err(to_js_error)?;
+        Ok(WasmWorker(Arc::new(sxg_worker)))
```

Comment:
Oh wow, so wasm-bindgen knows to serialize a handle to the `Arc` rather than its contents? How does this work? (Link to docs?)

Reply:
Oh OK, then why make it `Arc`? Why not a private `SxgWorker` field?

Reply:
This is due to the async member functions. For example, `WasmWorker::fetch_ocsp_from_ca(&self, fetcher)` is asynchronous, and `WasmWorker` is no longer a static variable. Hence it is possible that during the execution of `fetch_ocsp_from_ca()`, the `WasmWorker` instance is destructed by another thread. Wrapping `SxgWorker` into `Arc` extends the lifetime of `Arc<SxgWorker>` to the lifetime of the async function.